### DISCUSSION TITLE
Removing the dry submit of the Create API rule

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#6073] [Announcements]
Currently the Create API rule executes after a "dry submit" of the created record, or in other words - the Create API rule is executed AFTER the record is temporary inserted in the database within a transaction that is rollbacked at the end.

While this simplify some of the internals, it is not always intuitive for the end user and it often causes issues when self-referencing the collection we are trying to insert a record in, such as https://github.com/pocketbase/pocketbase/discussions/6071, https://github.com/pocketbase/pocketbase/discussions/5690, https://github.com/pocketbase/pocketbase/discussions/5551, https://github.com/pocketbase/pocketbase/discussions/5165 and others.

Therefore with the next v0.24.0 release I'm planning to change this behavior and execute the Create API rule without the "dry submit" so that it doesn't affect views and self-references. As a side-effect, it would also slightly improve the create request performance since there will be no dummy transaction to execute and rollback.

⚠️ **For most users this will not be a breaking change, BUT if you have relied on the existing behavior you will have to adjust manually your Create API rule to reflect the missing "dry submit".**

For example, in https://github.com/pocketbase/pocketbase/discussions/5690 since the user didn't want to use a UNIQUE index, the suggested workaround was to join a helper view collection that counts the inserted records and add `+1` for the "dry submit". With v0.24+ this no longer will be true. The fix in that case could be to discard the helper view collection and directly check against the self-referenced one similar to the initial user attempt, something like:
```js
invite != "" &&
(invite.email = "" || invite.email = email) &&
@collection.users.invite != invite
```

One drawback of this change could be potential misuse of the `@collection.*` field as in https://github.com/pocketbase/pocketbase/discussions/6071 where a UNIQUE index is usually the better and more performant solution. I'm not entirely sure how to make it more clear or properly explain it but I'll try to add a note in the documentation for the `@collection.*` field about this use case.

If you disagree with the change or have suggestions how to handle it better, feel free to let me know. In the meantime, I'll try to submit a PoC implementation in the `develop` branch ~a little bit later today~ _tomorrow_ for those who wants to test it.

**Comment by mirus-ua** [comment]
Hello. Thank you for the update. 
So, with these changes, we won't be able to check if a user creates an item with the correct relation.

Let's model the case: 
I have a table `customers`, and for each customer, I have an additional record in `customersMeta` that is related via customersMeta -> customers relation 
now I can do a check during the creation of a `customersMeta` like this
`@request.auth.id != "" && @request.auth.id = customer.user.id`, so I ensure that 

Now, do we need to cover these cases via hooks, or I oversee something?

**Reply by ganigeorgiev** [reply]
> we won't be able to check if a user creates an item with the correct relation.

I don't know how did you end up with this conclusion but it is incorrect.

The above change affects `@collection.*` self references and views that targets the SAME collection that the API rule is defined for (you can explore the linked discussions to see the related problems that it is trying to solve).

There is no change in the behavior for the rule: `@request.auth.id != "" && @request.auth.id = customer.user.id`.

**Reply by mirus-ua** [reply]
I definitely misread that. Thanks for your time and clarification

**Reply by ganigeorgiev** [reply]
If your confusion originate from the fact that the rule will execute before the creation and you assume that the record fields (aka. `customer`) will not be accessible/populated, then this is not the case.

The above change has no affect on the record fields resolver behavior and it works the same as it is right now, aka. for the Create API rule the main record fields like `customer.*` are still accessible and it is essentially an alias for `@request.body.customer.*`.

I have to do a little more testing but there is a PoC in the `develop` branch that you can play with against your local/non-production setup if you are still not sure whether the above will affect your collections.

**Comment by ganigeorgiev** [comment]
The Create API rule change is available in [v0.24.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.24.0).

If you need help with the migration of your affected Create API rules, feel free to open a Q&A discussion and I'll try to assist.

**Comment by newtechnology12** [comment]
hello i have proble of createting first superadmin

**Reply by mirus-ua** [reply]
In a server console type: pocketbase superuser create email@email password

**Reply by ganigeorgiev** [reply]
@newtechnology12 This has nothing to do with the above change (API rules don't apply for superusers).

Feel free to open a new Q&A discussion but please elaborate a little more on the problem and how you are running PocketBase to understand the issue.
